### PR TITLE
[SMTChecker] Fix ICE implicit conversion string literal -> byte

### LINKED ISSUE #9595: [SMTChecker] ICE because *arraySort->domain != *_index.sort
```
pragma experimental SMTChecker;
contract C {
  mapping(byte => int) x;
  constructor() {
    x[""];
  }
}
```

throws

https://github.com/ethereum/solidity/blob/241a564fcabb40b829b2ba497dd87d488b83fbb9/libsmtutil/SolverInterface.h#L135

### LINKED ISSUE #10086: SMTChecker chokes on constant bytes32 conversion
Minimal example:
```
pragma experimental SMTChecker;
contract SMT {
    bytes32 constant internal NULL_BYTES32 = bytes32('');
}
```

Running with `solc Version: 0.7.4+commit.3f05b770.Darwin.appleclang` yields:

```
Internal compiler error during compilation:
/Users/distiller/project/libsolidity/formal/SMTEncoder.cpp(884): Throw in function void solidity::frontend::SMTEncoder::visitTypeConversion(const solidity::frontend::FunctionCall &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what:
[solidity::util::tag_comment*] =
```

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/9595
Fixes https://github.com/ethereum/solidity/issues/10086

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -2240,6 +2243,16 @@ Expression const* SMTEncoder::leftmostBase(IndexAccess const& _indexAccess)
 	return base;
 }
 
+TypePointer SMTEncoder::keyType(TypePointer _type)
+{
+	if (auto const* mapType = dynamic_cast<MappingType const*>(_type))
+		return mapType->keyType();
```
Comment (hrkrshnn) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	if (auto const* mappingType = dynamic_cast<MappingType const*>(_type))
		return mappingType->keyType();
```
Minor comment. `mappingType` is more consistently used in the codebase.

**Review by hrkrshnn — DISMISSED** [review]
LGTM. A minor comment. A second review would be nice before merging.

File: libsolidity/formal/SMTEncoder.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Updated

**Review by hrkrshnn — DISMISSED** [review]

**Comment by bshastry** [comment]
The following test

```
pragma experimental SMTChecker;
contract C {
  function f(bytes calldata b) external {
    ((b[:])[5]);
  }
}
```

throws

https://github.com/ethereum/solidity/pull/10077/files#diff-25ffc7f23a998f16fd91366b10f609936dc12d46cec28f7e69b9d6a458fd665cR2253

**Comment by leonardoalt** [comment]
@bshastry on this PR?

**Comment by leonardoalt** [comment]
Yep, confirm.

**Comment by leonardoalt** [comment]
Updated

**Comment by leonardoalt** [comment]
Ready for review

**Comment by chriseth** [comment]
conflict.

**Comment by leonardoalt** [comment]
Rebased

**Review by chriseth — APPROVED** [review]

**Pull request merged**